documentation/cadl-structure-guidelines.md

```diff
@@ -0,0 +1,127 @@
+# Repository Guidelines for Cadl Specs
+
+## Purpose
+
+We need to formulate a strategy for checking in actual Cadl service specs for Azure. Service teams need to know where
+and how to check these in and tooling needs to know how to find and consume them. While OpenAPI specs traditionally have been just spec
+documents, Cadl "specs" could include service-specific functions, decorators and so forth that elevate them to the level of a
+library, and to maximize the benefits of Cadl, this approach needs to be supported as a first-class citizen.
+
+## Repository
+
+Cadl can co-exist with Swagger within the existing `azure-rest-api-specs` repository. This approach will make it easier to generate Swagger artifacts without needing to sync between repos at the downside of having to live with any baggage associated with the repo as it ages. There are currently over 1,000 issues and 500 open PRs in this repo.
+
+## Structure Overview
+
+This proposal strives to align with [Azure SDK Repo Structure Guidelines](https://azure.github.io/azure-sdk/policies_repostructure.html).
+
+We eliminate the distinction between data-plane and resource-manager, treating all packages as siblings under the service family folder. This is what the SDKs do today. The legacy `data-plane` and `resource-manager` folders where Swagger specs live will remain.
+
+A given Cadl folder could represent any of the following scenarios:
+
+- an SDK and service spec (simplest, common for small services)
+- a service spec only
+- an SDK spec that aggregates many small service specs (one SDK to many microservices)
+- an SDK spec that extracts a portion of a large service spec (many SDKs to a single large service)
+- a shared library that represents neither a service spec nor an SDK
+
+Examples below:
+
+```
+-> specification
+   -> confidentialledger
+      -> ConfidentialLedger               (data-plane SDK + service)
+      -> data-plane                       (swagger)
+      -> resource-manager                 (swagger)
+   -> compute
+      -> Compute.Management               (mgmt SDK + service)
+      -> Compute.Management.Shared        (shared)
+      -> Compute.CloudService.Management  (service)
+      -> Compute.Diagnostic.Management    (service)
+      -> Compute.Disk.Management          (service)
+      -> Compute.Gallery.Management       (service)
+      -> Compute.Skus.Management          (service)
+      -> data-plane                       (swagger)
+      -> resource-manager                 (swagger)
```

Comment:
Why create directories under "compute" that replicate the structure within "resource-manager"? Expanding this example a bit, I think it would look like:
```
   -> compute
      -> Compute.Management               (mgmt SDK + service)
      -> Compute.Management.Shared        (shared)
      -> Compute.CloudService.Management  (service)
      -> Compute.Diagnostic.Management    (service)
      -> Compute.Disk.Management          (service)
      -> Compute.Gallery.Management       (service)
      -> Compute.Skus.Management          (service)
      -> resource-manager                 (swagger)
        -> Microsoft.Compute
           -> CloudserviceRP
           -> ComputeRP
           -> DiagnosticRP
           -> DiskRP
           -> GalleryRP
           -> Skus
           -> common
```

While the mapping above is relatively obvious, I fear that this opens the door for crazy structures like:

```
   -> widgets
      -> Widgets.Management               (mgmt SDK + service)
      -> Widgets.Cats.Management  (service)
      -> Widgets.Dogs.Management    (service)
      -> Widgets.Horse.Management          (service)
      -> Widgets.Mule.Management       (service)
      -> Widgets.Sheep.Management          (service)
      -> resource-manager                 (swagger)
        -> Microsoft.Widgets
           -> Chevy
           -> Ford
           -> Nissan
           -> Rivian
           -> Tesla
           -> Toyota
```

Reply:
As we chatted offline the goal is to move to a module where data-plane and resource-manager folders become publish only sections and the work happens in the folders directly in the service directory. We know things aren't perfect and there is some risk there as well but we need to make forward progress and we think this is the best balance with all the requirements.

Reply:
Borrowing idea from @heaths (a.k.a. the Git-friendly folder structure, which link can be found on the related issue), and given Cadl is designed to support multiple versions with decorators in single files, we could use a `dev/` root folder, at the same level with `specification/`, to place all Cadl related files. For example:
```
.
├── dev (or vNext?)
│   └── keyvault
│       ├── KeyVault.Certificates
│       ├── KeyVault.Keys
│       ├── KeyVault.Management
│       │   ├── cadl-project.yaml
│       │   ├── examples
│       │   │   ├── CreateKey-example.json
│       │   │   ├── DeleteKey-example.json
│       │   │   ├── GetKey-example.json
│       │   │   └── GetKeys-example.json
│       │   ├── main.cadl
│       │   ├── tests
│       │   └── package.json
│       └── KeyVault.Secrets
└── specification
    └── keyvault
        ├── data-plane
        │   ├── Microsoft.KeyVault
        │   │   ├── preview
        │   │   └── stable
        │   └── readme.md
        └── resource-manager
            ├── Microsoft.KeyVault
            │   ├── preview
            │   └── stable
            └── readme.md
```
In this case, all Cadl related files will be in the `dev/` folder, and it will be always used to define the latest or 'vNext' API specs (so we may name the folder `vNext/`). 
When publishing, Swagger will be generated to the `specification/` folder - all the existing tools and pipeline based on current folder structure won't be affected at all.

Reply:
We are definitely using some of the dev related things such as versioning in place instead of side-by-side but we don't want to have multiple root level folders as we want the new cadl projects to live inside of the shared service folder so things are closer together.